**Important note:** make sure you click on the eye icon above to refresh the view and load the correct image.

For every combination of sample attributes (based on samples marked for reporting in the sample layout) and for each output time point, functional impact scores are plotted as a heatmap.
The x axis corresponds to positions in the mutated protein sequence.
The y axis corresponds to the introduced amino acid residues.
The wild-type nucleotide sequence is excluded from this visualization (the reference, with scores around 0, corresponds to synonymous codons, excluding the wild-type nucleotide sequence).
Only high confidence variants are shown. For a heatmap with many gray squares, it might be necessary to adjust the read count threshold used to classify variants.
The wild-type protein sequence is indicated by black dots.