{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e04bb879-08ef-4738-8dc5-8bdf79fceb71",
   "metadata": {},
   "source": [
    "# Notebook to generate expected mutants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ed79853-da94-4e77-8174-8d7a6125f968",
   "metadata": {},
   "source": [
    "## Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bd762f42-cf59-4468-a02c-d0b4d5f47cef",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4495a95f-2d5e-4485-b20b-f1d64b87863e",
   "metadata": {},
   "source": [
    "## Specify paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5b1ebc9e-57cd-417c-aa7f-22847272ecb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note, I used to import config entries as \"params\" (i.e. listed in the params statement of the rule), but...\n",
    "# a bug led to params not being imported.. (snakemake object is imported but not the params attribute)\n",
    "# Therefore, now I will import all config entries directly from the snakemake object\n",
    "wtseq_path = (\n",
    "    snakemake.input.wtseqs  # Projet-specific file containing the wild-type sequences\n",
    ")\n",
    "codon_table_path = snakemake.config[\"codon\"][\n",
    "    \"table\"  # Projet-specific file containing the genetic code\n",
    "]\n",
    "codon_mode = snakemake.config[\"codon\"][\n",
    "    \"mode\"  # Project-specific parameter to specify which degenerate codons were introduced\n",
    "]\n",
    "output_file = snakemake.output[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "748ee94c-8a44-4965-8b13-caf347cc2ccb",
   "metadata": {},
   "source": [
    "## Import codon table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fa65d047-0c07-41d1-961e-ad33ffe3a16a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>codon</th>\n",
       "      <th>aminoacid</th>\n",
       "      <th>freq</th>\n",
       "      <th>number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>26.26</td>\n",
       "      <td>76999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "      <td>17.89</td>\n",
       "      <td>52459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "      <td>26.31</td>\n",
       "      <td>77131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  codon aminoacid   freq  number\n",
       "0   TTT         F  26.26   76999\n",
       "1   TTC         F  17.89   52459\n",
       "2   TTA         L  26.31   77131"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_table = pd.read_csv(codon_table_path, header=0)\n",
    "codon_table[\"codon\"] = codon_table[\"codon\"].str.upper()\n",
    "codon_table.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9ea59b40-84ca-449c-849f-696333ccda5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to dictionary\n",
    "codon_dic = dict(zip(codon_table[\"codon\"], codon_table[\"aminoacid\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f5d6fef-ad77-44fc-8b10-9096938ccba9",
   "metadata": {},
   "source": [
    "## Define and test functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e90b1a36-e347-4e6f-a935-049ba3c13420",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_alt_codons(seq, codon_dic, mode=\"NNN\"):\n",
    "    \"\"\"\n",
    "    Based on a DNA sequence, the function returns two lists:\n",
    "    1) A list containing all 0-based amino acid positions for the sequence\n",
    "    2) A list containing all possible alternative codons (other than WT codon) at the matching positions\n",
    "    For list 2, the mode defines which codons are acceptable: NNN by default, or NNK\n",
    "    Codons are fetched in the provided codon table (dictionary)\n",
    "    \"\"\"\n",
    "\n",
    "    if mode == \"NNN\":\n",
    "        alt = [x for x in codon_dic.keys()]\n",
    "    elif mode == \"NNK\":\n",
    "        alt = [x for x in codon_dic.keys() if x[2] in [\"G\", \"T\"]]\n",
    "    else:\n",
    "        print(\"Pleae specify a correct mode: either NNN or NNK\")\n",
    "\n",
    "    pos_l = []\n",
    "    var_l = []\n",
    "\n",
    "    for i in range(0, len(seq), 3):\n",
    "        list_var = [x for x in alt if x != seq[i : i + 3]]\n",
    "        pos_l.append(i // 3)  # 0-based position (aa)\n",
    "        var_l.append(list_var)  # list of possible codons other than WT\n",
    "\n",
    "    return pos_l, var_l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d50c8d10-a434-417c-8007-7695196cef33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "([0, 1, 2], [['TTT', 'TTC', 'TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG', 'ATT', 'ATC', 'ATA', 'ATG', 'GTT', 'GTC', 'GTA', 'GTG', 'TCC', 'TCA', 'TCG', 'CCT', 'CCC', 'CCA', 'CCG', 'ACT', 'ACC', 'ACA', 'ACG', 'GCT', 'GCC', 'GCA', 'GCG', 'TAT', 'TAC', 'TAA', 'TAG', 'CAT', 'CAC', 'CAA', 'CAG', 'AAT', 'AAC', 'AAA', 'AAG', 'GAT', 'GAC', 'GAA', 'GAG', 'TGT', 'TGC', 'TGA', 'TGG', 'CGT', 'CGC', 'CGA', 'CGG', 'AGT', 'AGC', 'AGA', 'AGG', 'GGT', 'GGC', 'GGA', 'GGG'], ['TTT', 'TTC', 'TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG', 'ATT', 'ATC', 'ATA', 'ATG', 'GTT', 'GTC', 'GTA', 'GTG', 'TCT', 'TCC', 'TCA', 'TCG', 'CCC', 'CCA', 'CCG', 'ACT', 'ACC', 'ACA', 'ACG', 'GCT', 'GCC', 'GCA', 'GCG', 'TAT', 'TAC', 'TAA', 'TAG', 'CAT', 'CAC', 'CAA', 'CAG', 'AAT', 'AAC', 'AAA', 'AAG', 'GAT', 'GAC', 'GAA', 'GAG', 'TGT', 'TGC', 'TGA', 'TGG', 'CGT', 'CGC', 'CGA', 'CGG', 'AGT', 'AGC', 'AGA', 'AGG', 'GGT', 'GGC', 'GGA', 'GGG'], ['TTT', 'TTC', 'TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG', 'ATT', 'ATC', 'ATA', 'ATG', 'GTC', 'GTA', 'GTG', 'TCT', 'TCC', 'TCA', 'TCG', 'CCT', 'CCC', 'CCA', 'CCG', 'ACT', 'ACC', 'ACA', 'ACG', 'GCT', 'GCC', 'GCA', 'GCG', 'TAT', 'TAC', 'TAA', 'TAG', 'CAT', 'CAC', 'CAA', 'CAG', 'AAT', 'AAC', 'AAA', 'AAG', 'GAT', 'GAC', 'GAA', 'GAG', 'TGT', 'TGC', 'TGA', 'TGG', 'CGT', 'CGC', 'CGA', 'CGG', 'AGT', 'AGC', 'AGA', 'AGG', 'GGT', 'GGC', 'GGA', 'GGG']])\n"
     ]
    }
   ],
   "source": [
    "print(get_alt_codons(\"TCTCCTGTT\", codon_dic, \"NNN\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d4d6ac23-7b5d-4026-9903-f15fcb078770",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nt_seq(seq, mut_dic):\n",
    "    list_codons = [\n",
    "        seq[i : i + 3] for i in range(0, len(seq), 3)\n",
    "    ]  # Convert nucleotide sequence to list of codons\n",
    "    seq_l = [\n",
    "        mut_dic[a] if a in mut_dic.keys() else wtcodon\n",
    "        for a, wtcodon in enumerate(list_codons)\n",
    "    ]\n",
    "    return \"\".join(seq_l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a707a663-cb9e-449c-84b7-4f332ac9ec71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TTCCCTTTA'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_nt_seq(\"TCTCCTGTT\", {0: \"TTC\", 2: \"TTA\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b8be913-8eec-4b30-b530-3b2beb8f200c",
   "metadata": {},
   "source": [
    "## Import sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "021524c8-32f1-47c3-a80f-7dc0ccf08353",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutated_seq</th>\n",
       "      <th>WT_seq</th>\n",
       "      <th>nt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Mutated_seq                                             WT_seq  \\\n",
       "0       CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "\n",
       "                                              nt_seq  \n",
       "0  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt_seq = pd.read_csv(wtseq_path)\n",
    "wt_seq[\"nt_seq\"] = wt_seq[\"WT_seq\"].str.upper()\n",
    "wt_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a87dce4-06bd-462a-bf6b-3b9c53f3a2a6",
   "metadata": {},
   "source": [
    "## Get mutated sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14b1cc53-e77e-49df-927b-b37298f97f22",
   "metadata": {},
   "source": [
    "### Get all allowed codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7e11128a-f321-4bfe-ae26-bd8d580fdd18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutated_seq</th>\n",
       "      <th>WT_seq</th>\n",
       "      <th>nt_seq</th>\n",
       "      <th>pos</th>\n",
       "      <th>alt_codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...</td>\n",
       "      <td>[[TTT, TTC, TTA, TTG, CTT, CTC, CTA, CTG, ATT,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Mutated_seq                                             WT_seq  \\\n",
       "0       CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "\n",
       "                                              nt_seq  \\\n",
       "0  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "\n",
       "                                                 pos  \\\n",
       "0  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...   \n",
       "\n",
       "                                          alt_codons  \n",
       "0  [[TTT, TTC, TTA, TTG, CTT, CTC, CTA, CTG, ATT,...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt_seq[\"pos\"], wt_seq[\"alt_codons\"] = zip(\n",
    "    *wt_seq.WT_seq.apply(lambda x: get_alt_codons(x, codon_dic, codon_mode))\n",
    ")\n",
    "wt_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fe4b96b-cf61-4204-b1bd-0b8b43398b0c",
   "metadata": {},
   "source": [
    "### Reshape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "da6a9281-b120-4ecb-8e5a-6123c57ebdce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutated_seq</th>\n",
       "      <th>WT_seq</th>\n",
       "      <th>nt_seq</th>\n",
       "      <th>pos</th>\n",
       "      <th>alt_codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>[TTT, TTC, TTA, TTG, CTT, CTC, CTA, CTG, ATT, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>1</td>\n",
       "      <td>[TTT, TTC, TTA, TTG, CTT, CTC, CTA, CTG, ATT, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Mutated_seq                                             WT_seq  \\\n",
       "0       CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0       CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "\n",
       "                                              nt_seq pos  \\\n",
       "0  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0   \n",
       "0  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   1   \n",
       "\n",
       "                                          alt_codons  \n",
       "0  [TTT, TTC, TTA, TTG, CTT, CTC, CTA, CTG, ATT, ...  \n",
       "0  [TTT, TTC, TTA, TTG, CTT, CTC, CTA, CTG, ATT, ...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "singles_compact = wt_seq.explode([\"pos\", \"alt_codons\"])\n",
    "singles_compact.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7e53e333-afe5-46c7-ae13-9561f56372f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutated_seq</th>\n",
       "      <th>WT_seq</th>\n",
       "      <th>nt_seq</th>\n",
       "      <th>pos</th>\n",
       "      <th>alt_codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>CTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>GGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>GGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>GGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4410 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mutated_seq                                             WT_seq  \\\n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "..         ...                                                ...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "\n",
       "                                               nt_seq pos alt_codons  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        TTT  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        TTC  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        TTA  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        TTG  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        CTT  \n",
       "..                                                ...  ..        ...  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        AGG  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        GGT  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        GGC  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        GGA  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        GGG  \n",
       "\n",
       "[4410 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "singles_df = singles_compact.explode(\"alt_codons\")\n",
    "singles_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "15566655-391b-4f2f-bebb-ff197dad0fe6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutated_seq</th>\n",
       "      <th>WT_seq</th>\n",
       "      <th>nt_seq</th>\n",
       "      <th>pos</th>\n",
       "      <th>alt_codons</th>\n",
       "      <th>mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTT</td>\n",
       "      <td>{0: 'TTT'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTC</td>\n",
       "      <td>{0: 'TTC'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTA</td>\n",
       "      <td>{0: 'TTA'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTG</td>\n",
       "      <td>{0: 'TTG'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>CTT</td>\n",
       "      <td>{0: 'CTT'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>AGG</td>\n",
       "      <td>{69: 'AGG'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>GGT</td>\n",
       "      <td>{69: 'GGT'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>GGC</td>\n",
       "      <td>{69: 'GGC'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>GGA</td>\n",
       "      <td>{69: 'GGA'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>GGG</td>\n",
       "      <td>{69: 'GGG'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4410 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mutated_seq                                             WT_seq  \\\n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "..         ...                                                ...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "\n",
       "                                               nt_seq pos alt_codons  \\\n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        TTT   \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        TTC   \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        TTA   \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        TTG   \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        CTT   \n",
       "..                                                ...  ..        ...   \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        AGG   \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        GGT   \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        GGC   \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        GGA   \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        GGG   \n",
       "\n",
       "      mutations  \n",
       "0    {0: 'TTT'}  \n",
       "0    {0: 'TTC'}  \n",
       "0    {0: 'TTA'}  \n",
       "0    {0: 'TTG'}  \n",
       "0    {0: 'CTT'}  \n",
       "..          ...  \n",
       "0   {69: 'AGG'}  \n",
       "0   {69: 'GGT'}  \n",
       "0   {69: 'GGC'}  \n",
       "0   {69: 'GGA'}  \n",
       "0   {69: 'GGG'}  \n",
       "\n",
       "[4410 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "singles_df[\"mutations\"] = singles_df.apply(\n",
    "    lambda row: {row[f\"pos\"]: row[f\"alt_codons\"]}, axis=1\n",
    ")\n",
    "singles_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a0ea873-2121-4ea2-9689-517be5644162",
   "metadata": {},
   "source": [
    "### Construct sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ad13260a-7caa-4900-adee-a136a43b829b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutated_seq</th>\n",
       "      <th>WT_seq</th>\n",
       "      <th>nt_seq</th>\n",
       "      <th>pos</th>\n",
       "      <th>alt_codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TTTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TTCCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TTACCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TTGCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>TTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>CTTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>0</td>\n",
       "      <td>CTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>AGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>GGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>GGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>GGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN_F1</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...</td>\n",
       "      <td>69</td>\n",
       "      <td>GGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4410 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mutated_seq                                             WT_seq  \\\n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "..         ...                                                ...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "0        CN_F1  TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   \n",
       "\n",
       "                                               nt_seq pos alt_codons  \n",
       "0   TTTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        TTT  \n",
       "0   TTCCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        TTC  \n",
       "0   TTACCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        TTA  \n",
       "0   TTGCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        TTG  \n",
       "0   CTTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...   0        CTT  \n",
       "..                                                ...  ..        ...  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        AGG  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        GGT  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        GGC  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        GGA  \n",
       "0   TCTCCTGTTGAAGGTTCTCCAGCTAAGCCAGAAGATTACCCACACT...  69        GGG  \n",
       "\n",
       "[4410 rows x 5 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "singles_df[\"nt_seq\"] = singles_df.apply(\n",
    "    lambda row: get_nt_seq(row.WT_seq, row.mutations), axis=1\n",
    ")\n",
    "singles_df.drop(columns=\"mutations\", inplace=True)\n",
    "singles_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0deb58f4-e279-4ea7-b093-6607e62a6ee0",
   "metadata": {},
   "source": [
    "## Export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c2c3a02-c8f5-4ec0-9228-ebd73ebaae95",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note: information on mutated codons is dropped so that we can annotate more types of mutants in downstream steps\n",
    "singles_df.drop(['pos','alt_codons'], axis=1).to_csv(output_file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
